### DISCUSSION TITLE
Memory leak from the local context into the global one

### DISCUSSION THREAD

**Discussion by baikinkirill** [mikro-orm/mikro-orm#5309] [General]
### Describe the bug

When using a transaction in a loop, in which only one entity is requested and updated, the data from this transaction still seeps from the transaction context into the global one, thereby clogging up the application memory.

### Reproduction

I have compiled a test repository with the problem reproduction and measurements, you can check it out at the link: https://github.com/baikinkirill/mikroorm-em-demo

There, I demonstrated the most simplified and trimmed down logic that we have in our application.

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.1.7

### Node.js version

20.10.0

### Operating system

macOS Sonoma 14.3.1 (23D60)

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
Thanks for the report, but this doesn't seem to be a problem in the ORM codebase to me. If you think its a memory leak, please provide an isolated reproduction without any web servers and other unnecessary dependencies - this needs to be as simple as possible, a single script ideally, meaning also no stuff like `p-map` to rule out possible issues in other libraries. 

> When using a transaction in a loop, in which only one entity is requested and updated, the data from this transaction still seeps from the transaction context into the global one, thereby clogging up the application memory.

This feels like you are describing the expected and correct behavior of explicit transactions - the changes they make are always propagated to the upper level, and the upper level is not supposed to be global.

Note that this particular part of your repro seems completely off (and I guess that is what results in your memory leak). `this.#em` is a single EM instance (and based on what you are saying, its global, which is wrong on its own), and you are trying to fire multiple promises in parallel while calling `this.#em.clear()`. This is destined to fail even if without a memory leak, it kinda confirms you have a global context in `this.#em`, which, again, is wrong. You are not supposed to call `em.clear()` in your code base (unless in a few very specific cases like huge batch inserts), this is a helper for tests or for recovery (when flush causes error, so you can clear and start over).

https://github.com/baikinkirill/mikroorm-em-demo/blob/master/src/app.service.ts#L23-L34

And a final note, this code is super unperformant. You want to run everything in a single context/single flush, that way you get a siglne update query per table. Just the fact that you load N entities with N findOne calls is wrong (the good old N+1 problem), and with your implementation, you do the same thing for updating. So instead of 1 select and 1 update query for all your entities, you do 2*N queries.

**Reply by B4nan** [reply]
I guess we could skip the propagation to the upper level if it's a global one, that would probably fix your problem on its own. But it doesn't change the fact that your `this.#em` shouldn't be global, it should be already request scoped, so you shouldn't fall into this at all. Also, it feels a bit breaking, you never know what behaviors were people depending on (edit: but we could safely do it if the global context is not allowed, which is the default).

What do you get if you `console.log(this.#em.id)` before you create the transaction?

**Reply by B4nan** [reply]
Maybe this will help, let me know:

https://github.com/mikro-orm/mikro-orm/commit/7ac9a1945469264ea32727c2e63fd89d9f2d7ec8?w=1

If not, it means your `this.#em` is not global, which I don't see how could happen in your repro. 

But again, if it was global, it was wrong in the first place, and it points to a broken async context most probably.

**Reply by baikinkirill** [reply]
The thing is, we have a background task service (bgtt) that's designed for executing complex and lengthy operations without overloading the main API pods. This logic of searching for and updating entities individually in a loop is necessary for us to avoid locking all user entities. Therefore, instead of locking everything, we lock and update them one by one.

Yes, one such operation in bgtt is executed in a single RequestContext.

> What do you get if you console.log(this.#em.id) before you create the transaction?

It prints out the identifier 4.

> This feels like you are describing the expected and correct behavior of explicit transactions - the changes they make are always propagated to the upper level, and the upper level is not supposed to be global.

Why do you think the data from the local transaction context should seep into the higher level?

> Maybe this will help, let me know: https://github.com/mikro-orm/mikro-orm/commit/7ac9a1945469264ea32727c2e63fd89d9f2d7ec8?w=1

Thank you for the suggestions, I'll try them now and get back to you.

**Reply by baikinkirill** [reply]
Such changes didn't help; entities from the transaction context are still seeping into the higher level. Perhaps I initially made a mistake, and I should have mentioned not the global context but the request context.

**Reply by B4nan** [reply]
Yeah if it wasn't a global context then this can't help, and it is a wontfix to me, you should fork instead of clearing: 

```ts
async updateEntities(ids: number[]) {
  await pMap(
    ids,
    async (id: number) => {
      await this.#em.fork().transactional(async (em) => {
        await this.#updateEntity(id);
      });
    },
    { concurrency: 3 },
  );
}
```

> Why do you think the data from the local transaction context should seep into the higher level?

It was always working like that, just like `em.transaction` copies everything from the outer `em` context into it (so you can work with entities you already loaded before), it puts the changes done by the callback back once it finishes. The idea is that it represents `begin/commit`, not that it creates some isolated side-effect-free context - that is what `em.fork()` is for.

**Reply by B4nan** [reply]
And I am not entirely sure if this on its own will fix things, as the idea of having non-global EM that is gonna be reused stored in `this.#em` seems just wrong. Your private `updateEntity` method is using it, and since its not global, it probably won't respect the context created by `em.transactional`, only the global EM does that.

**Reply by B4nan** [reply]
> And I am not entirely sure if this on its own will fix things, as the idea of having non-global EM that is gonna be reused stored in this.#em seems just wrong. Your private updateEntity method is using it, and since its not global, it probably won't respect the context created by em.transactional, only the global EM does that.

But you can fork with `useContext: true` to make the fork respect contexts automatically, just like the global one does to get around that. Note that this would need to be used when creating the `this.#em` value (which seems like it should came from nest DI so it should be the global one, but if you say its not...)

**Reply by baikinkirill** [reply]
Calling `this.#em.fork()` with `useContext: true`, according to the documentation, means that the request context will be utilized. I tried it, and all entities utilized in the transaction remained in memory even after the transaction had concluded, so this option did not solve my problem.

However, `this.#em.fork().transactional()` did help, as there were at most five entities in memory simultaneously (although I have not yet tested this with several hundred entities). Hence, the problem of entity retention and memory overflow on the pod is likely to be resolved.

Initially, I expected that nothing should seep out from the transaction after its completion.

> And I am not entirely sure if this on its own will fix things, as the idea of having non-global EM that is gonna be reused stored in this.#em seems just wrong. Your private updateEntity method is using it, and since its not global, it probably won't respect the context created by em.transactional, only the global EM does that.

So, do I understand correctly that if I continue to use `this.#em` inside the transaction, it will refer to the request context, not the transaction context?

**Reply by B4nan** [reply]
> So, do I understand correctly that if I continue to use this.#em inside the transaction, it will refer to the request context, not the transaction context?

If it's not a global EM instance (and not a fork created with `useContext: true`), this means it will not respect any context, let it be the one from `RequestContext` helper (created by a middleware automatically) or the one that is created via `em.transactional`. Such forks always work with their own identity map.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/EntityManager.ts#L1907

If you want to use dependency injection, you need to work with a fork that respects the context.
